### PR TITLE
password and logOut functionalities

### PR DESCRIPTION
**Pull request by pereebi**
Forgot password, reset password, and logout functionalities done.
Please vet.

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,23 @@
+P0RT=442
```
Comment (forestbaba) [inline .env:1]: Why are you adding .env to your git??....This shouldn't happen remove it

File: .env [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+ENV=dev
+
+GMAIL_HOST=mail.creditalert.com.ng
+GMAIL_PORT=465
+GMAIL_USESSL=true
+GMAIL_USERNAME=info@creditalert.com.ng
+GMAIL_PASSWORD=tx9fOpO?PExG
+FROM=Samtechnet Team
+FRO='"Squazzle Team" <info@creditalert.com.ng>'
```
Comment (forestbaba) [inline .env:18]: What's this line ?

File: dist/mailers/senders.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
+"use strict";
```
Comment (forestbaba) [inline dist/mailers/senders.js:1]: Remove build folder from PR...Add it to git ignore

File: dist/middleware/errrorController.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
+"use strict";
```
Comment (forestbaba) [inline dist/middleware/errrorController.js:1]: Same here

File: dist/middleware/validateRequest.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
+"use strict";
```
Comment (forestbaba) [inline dist/middleware/validateRequest.js:1]: Here too

File: dist/middleware/validation.js [hunk 6]
```diff
@@ -0,0 +1,269 @@
+"use strict";
```
Comment (forestbaba) [inline dist/middleware/validation.js:1]: Same here

File: dist/model/user.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
+"use strict";
```
Comment (forestbaba) [inline dist/model/user.js:1]: Same here

File: dist/repositories/user-repository.js [hunk 8]
```diff
@@ -0,0 +1,317 @@
+"use strict";
```
Comment (forestbaba) [inline dist/repositories/user-repository.js:1]: Same here

File: dist/routes/userRoutes.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+"use strict";
```
Comment (forestbaba) [inline dist/routes/userRoutes.js:1]: Same here

File: dist/server.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
+"use strict";
```
Comment (forestbaba) [inline dist/server.js:1]: Same here

File: dist/services/errorHandlers/errors.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
+"use strict";
```
Comment (forestbaba) [inline dist/services/errorHandlers/errors.js:1]: Same here...Remove everything that has to do with dist in git commit

File: src/controller/user.ts [hunk 12]
```diff
@@ -408,6 +408,115 @@ const changePassword = async (
...
+    }
+
+    const userInformation = await UserData.findOneUser(email);
+
+    await UserData.updateUserVerificationCode(userInformation.emails, passwordCode);
+
+    const convert = compile({
+      wordwrap: 130,
+    });
+    const htmls = [
```
Comment (forestbaba) [inline src/controller/user.ts:432]: Abstract this to a separate function in like utility file or something then import...This makes the code messy

File: src/controller/user.ts [hunk 13]
```diff
@@ -408,6 +408,115 @@ const changePassword = async (
...
+    const texts = htmls.map(convert);
+    const mailOptions = {
+      from: FRO,
+      to: email,
+      text: `Hello ${userInformation.firstName}`,
+
+      subject: "Forgot Password",
+      html: texts.join(" \n"),
+    };
+    sendMail.send(mailOptions);
```
Comment (forestbaba) [inline src/controller/user.ts:448]: Just call send mail here, everything that has to do with the details of the mail like above isn't needed

File: src/controller/user.ts [hunk 14]
```diff
@@ -408,6 +408,115 @@ const changePassword = async (
...
+    const verifyUserData = await UserData.findUserByCode(code);
+
+    if (verifyUserData == "isNotValid") {
+      return res.status(Statuscode.unauthorized()).json({
+        message: "Invalid code",
+        success: false,
+      });
+    }
+
+    if (password != confirmPassword) {
```
Comment (forestbaba) [inline src/controller/user.ts:474]: Use deep equals here

File: src/controller/user.ts [hunk 15]
```diff
@@ -408,6 +408,115 @@ const changePassword = async (
...
+    );
+  }
+};
+
+const resetPassword = async (req: Request, res: Response, next: NextFunction) => {
+  const { code, password, confirmPassword } = req.body;
+  try {
+    const verifyUserData = await UserData.findUserByCode(code);
+
+    if (verifyUserData == "isNotValid") {
```
Comment (forestbaba) [inline src/controller/user.ts:467]: Use deep equals

File: src/controller/user.ts [hunk 16]
```diff
@@ -408,6 +408,115 @@ const changePassword = async (
...
+    return next(
+      new AppError(`something went wrong here is the error ${error}`, Statuscode.internalServerError())
+    );
+  }
+};
+
+const logOut =async (req:Request, res: Response, next:NextFunction) => {
+  try {
+    const authorization = req.headers.authorization;    
+    const token = (authorization || "").split(" ")[1];
```
Comment (forestbaba) [inline src/controller/user.ts:499]: Ask samuel about the refactoring here....U don't actually need like 499 to 501

**Comment by pereebi** [comment]
corrections have been implemented

**Review by forestbaba — COMMENTED** [review]
The dist folder is still in git commit, remove it

**Pull request closed without merging**